### PR TITLE
Blofin implementation

### PR DESCRIPTION
**Pull request by luckydss**
Exchange Name
BloFin

URL of the Exchange
https://blofin.com/

URL of the Exchange's API docs
https://blofin.com/docs

Why should this exchange be implemented?
Blofin is a rapidly growing exchange that provides users with fast and reliable trading as well as secure asset management.

### CODE REVIEW

**Review by pcriadoperez — COMMENTED** [review]
Thanks a lot for contributing @luckydss .

However please review the contributing docs. 
Most of the code we write in typescript following certain rules, and then it's automatically transpiled to other languages. So we look to only commit the .ts files

https://docs.ccxt.com/#/CONTRIBUTING

**Comment by luckydss** [comment]
> Thanks a lot for contributing @luckydss .
> 
> However please review the contributing docs. Most of the code we write in typescript following certain rules, and then it's automatically transpiled to other languages. So we look to only commit the .ts files
> 
> https://docs.ccxt.com/#/CONTRIBUTING

Thank you for your reply. Your suggestion is received. We are developing a typescript version and expect to submit it this week.

**Comment by luckydss** [comment]
> Thanks a lot for contributing @luckydss .
> 
> However please review the contributing docs. Most of the code we write in typescript following certain rules, and then it's automatically transpiled to other languages. So we look to only commit the .ts files

Hello, thank you again for your suggestions. Then according to your suggestions, we made modifications and added typescript code. Please check again and look forward to your reply.

> > Thanks a lot for contributing @luckydss .
> > However please review the contributing docs. Most of the code we write in typescript following certain rules, and then it's automatically transpiled to other languages. So we look to only commit the .ts files
> > https://docs.ccxt.com/#/CONTRIBUTING
> 
> Thank you for your reply. Your suggestion is received. We are developing a typescript version and expect to submit it this week.

Hello, thank you again for your suggestions. According to your suggestions, we have made modifications and added typescript code. Please check again and look forward to your reply.

**Comment by carlosmiei** [comment]
@luckydss thanks, we will review it soon

File: ts/src/blofin.ts [hunk 1]
```diff
@@ -0,0 +1,2165 @@
...
+                return markets[0];
+            } else if ((symbol.indexOf ('-C') > -1) || (symbol.indexOf ('-P') > -1)) {
+                return this.createExpiredOptionMarket (symbol);
+            }
+        }
+        throw new BadSymbol (this.id + ' does not have market symbol ' + symbol);
+    }
+
+    safeMarket (marketId = undefined, market = undefined, delimiter = undefined, marketType = undefined) {
+        const isOption = (marketId !== undefined) && ((marketId.indexOf ('-C') > -1) || (marketId.indexOf ('-P') > -1));
```
Comment (carlosmiei) [inline ts/src/blofin.ts:436]: does blofin support options? it does not seem so,  i think we can remove this method

Comment (luckydss) [reply ts/src/blofin.ts]: Thank you very much for your review, and thank you for the code modifications you made. Yes, your suggestions were absolutely correct, and we have made the necessary changes and updates. We look forward to your review once again.

Comment (carlosmiei) [reply ts/src/blofin.ts]: @luckydss no problem, there are still some fixes pending, but we will handle it, no worries

Comment (luckydss) [reply ts/src/blofin.ts]: Great, thank you for letting me know. If there is anything I can do to help, please don't hesitate to reach out. Have a great day!

Comment (carlosmiei) [reply ts/src/blofin.ts]: @luckydss Sure! You too!

**Comment by carlosmiei** [comment]
blofin-resized

**Comment by carlosmiei** [comment]
blofin-white-resized

**Comment by carlosmiei** [comment]
blofin-new-white-resized

**Pull request merged**